src/Physics/HadronTransport/Intranuke2025.cxx

```diff
@@ -0,0 +1,406 @@
+
+       // if I can't rescatter it, I will just take it out of the nucleus
+       LOG("Intranuke2025", pNOTICE)
+              << "... Current version can't rescatter a " << sp->Name();
+       sp->SetFirstMother(icurr);
+       sp->SetStatus(kIStStableFinalState);
+       evrec->AddParticle(*sp);
+       delete sp;
+       continue; // <-- skip to next GHEP entry
+    }
+
+    // Start stepping particle out of the nucleus
+    bool has_interacted = false;
+    while ( this-> IsInNucleus(sp) )
+    {
+      // advance the hadron by a step
+      utils::intranuke2025::StepParticle(sp, fHadStep);
+
+      // check whether it interacts
+      double d = this->GenerateStep(evrec,sp);
+      has_interacted = (d<fHadStep);
+      if(has_interacted) break;
+    }//stepping
+
+    if(has_interacted && fRemnA>0)  {
+        // the particle interacts - simulate the hadronic interaction
+      LOG("Intranuke2025", pNOTICE)
+          << "Particle has interacted at location:  "
+          << sp->X4()->Vect().Mag() << " / nucl rad= " << fTrackingRadius;
+	this->SimulateHadronicFinalState(evrec,sp);
+    } else if(has_interacted && fRemnA<=0) {
+        // nothing left to interact with!
+      LOG("Intranuke2025", pNOTICE)
+          << "*** Nothing left to interact with, escaping.";
+	sp->SetStatus(kIStStableFinalState);
+	evrec->AddParticle(*sp);
+	evrec->Particle(sp->FirstMother())->SetRescatterCode(1);
+    } else {
+        // the exits the nucleus without interacting - Done with it!
+        LOG("Intranuke2025", pNOTICE)
+          << "*** Hadron escaped the nucleus! Done with it.";
+	sp->SetStatus(kIStStableFinalState);
+	evrec->AddParticle(*sp);
+	evrec->Particle(sp->FirstMother())->SetRescatterCode(1);
+    }
+    delete sp;
+
+    // Current snapshot
+    //LOG("Intranuke2025", pINFO) << "Current event record snapshot: " << *evrec;
+
+  }// GHEP entries
+
+  // Add remnant nucleus - that 'hadronic blob' has all the remaining hadronic
+  // 4p not  put explicitly into the simulated particles
+  TLorentzVector v4(0.,0.,0.,0.);
+  GHepParticle remnant_nucleus(
+    kPdgHadronicBlob, kIStFinalStateNuclearRemnant, inucl,-1,-1,-1, fRemnP4, v4);
+  evrec->AddParticle(remnant_nucleus);
+  // Mark the initial remnant nucleus as an intermediate state
+  // Don't do that in the hadron/photon-nucleus scatterig mode since the initial
+  // remnant nucleus and the target nucleus coincide.
+  if(fGMode != kGMdHadronNucleus &&
+     fGMode != kGMdPhotonNucleus) {
+     evrec->Particle(inucl)->SetStatus(kIStIntermediateState);
+  }
+}
+//___________________________________________________________________________
+double Intranuke2025::GenerateStep(GHepRecord*  /*evrec*/, GHepParticle* p) const //Added ev to get tgt argument//
+{
+// Generate a step (in fermis) for particle p in the input event.
+// Computes the mean free path L and generate an 'interaction' distance d
+// from an exp(-d/L) distribution
+
+  int pdgc = p->Pdg();
+
+  double scale = 1.;
```

Comment:
The following set of if/then/else doesn't handle `kPdgKP` which is one of the allowed `CanRescatter()` codes.   This implies for that case the `scale` is `1`.   Is that sensible?  Should that case be called out explicitly?